feat: multiple accounts and microsoft provider

iboughtbed opened the pull request:
1. packages/db

- added email, name, and image fields into account schema

2. packages/auth

- linking multiple accounts
- created a database hook to retrieve the account's email, name, and image from the provider

3. packages/api

- created type/interfaces for provider classes (/providers/types.ts): it includes the interface that implements the calendar provider classes
- created microsoft calendar provider (graph-client), modified google calendar provider
- CRUD for events and retrieving calendars in each provider

- created a trpc procedure `calendarProcedure` that fetches all connected accounts of the user
- created functions to retrieve user's connections (/utils/accounts.ts)
- on each `getAllAccounts` call, we have to get the access token from better-auth in order to have fresh tokens if the access token expires

4. apps/web

- sticky sidebar calendar picker
- scroll only on calendars list in sidebar
- created a dialog to link an account
- updated types on frontend to match the procedure output through trpc router infer type

---

changed event.start and event.end structure, but this later will be resolver with the new Temporal PR

mezotv commented on bun.lock:
```diff
@@ -77,6 +77,8 @@
       "version": "0.1.0",
       "dependencies": {
+        "@microsoft/microsoft-graph-client": "^3.0.7",
+        "@microsoft/microsoft-graph-types": "^2.40.0",
```
If this is just types wouldn't it make more sense to just make it a dev dependency?

iboughtbed replied:
perhaps, but it's not a priority rn, gonna check it when finishing the PR ;)

a-rebets commented on packages/api/src/providers/microsoft-calendar.ts:
As per last PR, we better not add inline SVGs. If you need brand icons you can use:

https://github.com/simple-icons/simple-icons

iboughtbed commented:
1. packages/db

- created `connections` table which has user, account and provider (google/microsoft) id's.
- each connections has their own access tokens for their providers

2. packages/auth

- included microsoft social provider
- setup proper scopes for microsoft and google to receive refresh tokens (`offline` and `prompt: "consent"`)
- passed additional fields to user object in session (`defaultConnectionId`)
- created database hooks for account create/update that upserts the connection for the account
- created providers (a.k.a drivers) for google and microsoft to get the user's information (name, email, picture)

3. packages/api

- created type/interfaces for provider classes (/providers/types.ts): it includes the interface that implements the calendar provider classes
- created microsoft calendar provider (graph-client), modified google calendar provider
- CRUD for events and retrieving calendars in each provider

- created a trpc procedure `calendarProcedure` that fetches all connections of the user
- created functions to retrieve user's connections (/utils/connection.ts)
- on each `getAllConnections` call, we have to get the access token from better-auth in order to have fresh tokens if the access token expires

4. apps/web

- sticky sidebar calendar picker
- scroll only on calendars list in sidebar
- created a dialog to link an account
- updated types on frontend to match the procedure output through trpc router infer type

---

changed event.start and event.end structure, but this later will be resolver with the new Temporal PR

iboughtbed commented on packages/auth/src/providers/google.ts:
```diff
@@ -0,0 +1,63 @@
+
+  async getUserInfo() {
+    try {
+      const response = await this.people.people.get({
```
Yep, I found it:

https://github.com/better-auth/better-auth/blob/main/packages/better-auth/src/api/routes/account.ts#L515

iboughtbed replied:
It's introduced in one of the latest Beta versions, I'll update it to `1.2.9-beta.8`, all good?

iboughtbed commented:
@JeanMeijer @a-rebets , should I merge?

JeanMeijer commented on packages/db/drizzle/0005_careful_union_jack.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE "account" ALTER COLUMN "email" SET DEFAULT '';
```
You need to recreate these migrations (preferably into a single one). 

They currently fail when I try apply to a database with the schema that existed before this PR.

JeanMeijer commented on packages/db/src/schema/auth.ts:
```diff
@@ -28,7 +29,12 @@ export const session = pgTable("session", {
   id: text().primaryKey(),
   accountId: text().notNull(),
-  providerId: text().notNull(),
+  providerId: text().$type<"google" | "microsoft">().notNull(),
```
Prefer using [`text({ enum: ["google", "microsoft"] })`](https://orm.drizzle.team/docs/column-types/pg#text) which is safer than overwriting the types entirely.

iboughtbed replied:
nicee, didn't know this exists

JeanMeijer commented on packages/db/src/schema/auth.ts:
```diff
@@ -28,7 +29,12 @@ export const session = pgTable("session", {
+  // we are using $type<string> because during account creation
+  // better-auth does not pass additional fields
+  name: text().$type<string>().notNull().default(""),
+  email: text().$type<string>().notNull().default(""),
```
What do you mean with this? 

What is `$type<string>` supposed to do?
Seems like you're overriding the type `string` with `string`.

Also prefer using `null` as a value to indicate the data is missing.

iboughtbed replied:
database hooks pass email/name/image right after the account was created, so I default to empty string, because we might have type errors in the app when using account.email/image/name

JeanMeijer commented on packages/api/src/utils/accounts.ts:
```diff
@@ -0,0 +1,77 @@
+
+    const accounts = await Promise.all(
+      _accounts.map(async (account) => {
+        console.log("fetching...", account.id, account.providerId, user.id);
```
Can you remove this?

JeanMeijer commented on packages/api/src/trpc.ts:
```diff
@@ -52,3 +55,38 @@ export const protectedProcedure = t.procedure.use(({ ctx, next }) => {
+      (account) => account.id === ctx.user.defaultAccountId,
+    );
+
+    if (!activeAccount) {
+      throw new TRPCError({
+        code: "PRECONDITION_FAILED",
+        message: "No active account set",
+      });
+    }
```
The only thing I see when I run dev. 

You're filtering out other errors in `getAllAccounts`.
Consequently, we only discover something is wrong here, and now we don't know what they were and how to handle them.

JeanMeijer commented on packages/api/src/routers/accounts.ts:
```diff
@@ -0,0 +1,64 @@
+  delete: protectedProcedure
+    .input(z.object({ accountId: z.string() }))
+    .mutation(async ({ ctx, input }) => {
+      await ctx.db
+        .delete(account)
+        .where(
+          and(
+            eq(account.accountId, input.accountId),
+            eq(account.userId, ctx.user.id),
+          ),
+        );
+
+      const activeAccount = await getActiveAccount(ctx.user, ctx.headers);
+      if (activeAccount.accountId === input.accountId) {
+        await ctx.db
+          .update(user)
+          .set({ defaultAccountId: null })
+          .where(eq(user.id, ctx.user.id));
+      }
+    }),
```
It's possible for users to delete their only linked account, causing them to lose access.

Prefer instead using using `unlinkAccount` from [Better-Auth client](https://www.better-auth.com/docs/concepts/users-accounts#account-unlinking).

JeanMeijer commented on apps/web/src/components/add-account-dialog.tsx:
```diff
@@ -0,0 +1,72 @@
+      });
+      setOpen(false);
+    } catch (error) {
+      console.error("Failed to link account:", error);
```
Prefer using toast here.

JeanMeijer commented on packages/api/src/routers/events.ts:
```diff
@@ -17,159 +14,165 @@ export const eventsRouter = createTRPCRouter({
+            try {
+              const calendars = await client.calendars();
+              calendarIds = calendars
+                // .filter(
```
Don't forget to explain why you left this commented out for future readers.

JeanMeijer commented on packages/api/src/utils/date-helpers.ts:
```diff
@@ -15,15 +17,53 @@ export const dateHelpers = {
 
   parseGoogleDate(
-    googleDate: { date?: string; dateTime?: string },
+    googleDate: { date?: string; dateTime?: string; timeZone?: string },
```
Nice! Adding the timeZone field everywhere saves us time transitioning to Temporal later.

JeanMeijer commented on packages/auth/package.json:
```diff
@@ -10,18 +10,22 @@
   },
   "dependencies": {
+    "@googleapis/people": "^3.0.9",
+    "@microsoft/microsoft-graph-client": "^3.0.7",
     "@repo/db": "workspace:*",
-    "@repo/env": "workspace:*"
+    "@repo/env": "workspace:*",
+    "google-auth-library": "^9.15.1"
```
Don't think we need `@googleapis/people`, or `google-auth-library` anymore, right?

JeanMeijer reviewed (changes requested):
There are multiple type errors that are preventing the build from completing.

You can find them e.g. by running `bun run build`. 

Always run the build command, the dev server doesn't check the entire project or perform the same actions. I forget to do this too, and things break more often than you would think.
